src/rust/engine/hashing/src/lib.rs

```diff
@@ -312,6 +313,63 @@ impl<W: Write> Write for WriterHasher<W> {
   }
 }
 
+///
+/// Copy the data from reader and hash the bytes in one pass.
+/// Use hash() to just hash without copying the data anywhere.
+///
+pub fn copy_and_hash<R: ?Sized, W: ?Sized>(reader: &mut R, writer: &mut W) -> io::Result<Digest>
+where
+  R: io::Read,
+  W: io::Write,
+{
+  let mut hasher = WriterHasher::new(writer);
+  let _ = io::copy(reader, &mut hasher)?;
+  Ok(hasher.finish().0)
+}
+
+///
+/// Hashes the data into a Digest.
+///
+pub fn hash<R: ?Sized>(reader: &mut R) -> io::Result<Digest>
+where
+  R: io::Read,
+{
+  copy_and_hash(reader, &mut io::sink())
+}
+
+///
+/// Hashes the data into a Digest.
+///
+pub fn hash_path<P: AsRef<Path> + std::fmt::Debug + Clone>(path: P) -> Result<Digest, String> {
+  copy_and_hash(
+    &mut std::fs::File::open(path.clone()).map_err(|e| format!("Failed to open {path:?}: {e}."))?,
+    &mut io::sink(),
+  )
+  .map_err(|e| format!("Failed to hash {path:?}: {e}."))
+}
+
+///
+/// Copy from reader to writer and return whether the copied data matches expected_digest.
+///
```

Comment:
The utility functions lifted from `sharded_lmdb` are synchronous (because we `spawn_blocking` all interaction with the LMDB store), and so are unlikely to be usable outside of that context (... without wrapping in `spawn_blocking`). You'll encounter that headache when find-replacing to `tokio::fs`, unfortunately.

Reply:
oh no... I've encountered it